test/test_c10d_spawn.py

```diff
@@ -0,0 +1,189 @@
+    @unittest.skipIf(not TEST_MULTIGPU, "At least 2 CUDA GPUS needed")
+    @unittest.skipIf(NO_NCCL, "NCCL needed")
+    def test_shared_allreduce_nccl(self):
+        with tempfile.NamedTemporaryFile(delete=False) as file:
+            shared_tensors = [torch.ones(2, 2).to(i) for i in range(2)]
+            mp.spawn(ProcessGroupShareTensorTest._test_allreduce_process,
+                     args=(file.name,
+                           shared_tensors,
+                           self.world_size,
+                           ProcessGroupShareTensorTest._init_pg_nccl),
+                     nprocs=self.world_size,
+                     join=True)
+
+    @classmethod
+    def _test_reduce_process(
+            cls, rank, filename, shared_tensors, world_size, init_pg):
+        pg = init_pg(rank, filename, world_size)
+        x = shared_tensors[rank]
+        pg.reduce(x, root=0, op=c10d.ReduceOp.SUM).wait()
+        if rank == 0:
+            cls.assert_equal(torch.ones(2, 2) * 2, x.to("cpu"))
+        else:
+            cls.assert_equal(torch.ones(2, 2), x.to("cpu"))
+
+    @unittest.skipIf(not TEST_MULTIGPU, "At least 2 CUDA GPUS needed")
+    @unittest.skipIf(NO_NCCL, "NCCL needed")
+    def test_shared_reduce_nccl(self):
+        with tempfile.NamedTemporaryFile(delete=False) as file:
+            shared_tensors = [torch.ones(2, 2).to(i) for i in range(2)]
+            mp.spawn(ProcessGroupShareTensorTest._test_reduce_process,
+                     args=(file.name,
+                           shared_tensors,
+                           self.world_size,
+                           ProcessGroupShareTensorTest._init_pg_nccl),
+                     nprocs=self.world_size,
+                     join=True)
+
+    @classmethod
+    def _test_allgather_process(
+            cls, rank, filename, shared_tensors, world_size, init_pg):
+        pg = init_pg(rank, filename, world_size)
+        xs = [shared_tensors[rank]]
+        ys = [[torch.zeros_like(xs[0]) for i in range(world_size)]]
+        pg.allgather(ys, xs).wait()
+        for i in range(world_size):
+            cls.assert_equal(torch.ones(2, 2) * i, ys[0][i].to("cpu"))
+
+    @unittest.skipIf(not TEST_MULTIGPU, "At least 2 CUDA GPUS needed")
+    def test_shared_allgather_gloo(self):
+        with tempfile.NamedTemporaryFile(delete=False) as file:
+            shared_tensors = [torch.ones(2, 2).to(i) * i for i in range(2)]
+            mp.spawn(ProcessGroupShareTensorTest._test_allgather_process,
+                     args=(file.name,
+                           shared_tensors,
+                           self.world_size,
+                           ProcessGroupShareTensorTest._init_pg_gloo),
+                     nprocs=self.world_size,
+                     join=True)
+
+    @unittest.skipIf(not TEST_MULTIGPU, "At least 2 CUDA GPUS needed")
+    @unittest.skipIf(NO_NCCL, "NCCL needed")
+    def test_shared_allgather_nccl(self):
+        with tempfile.NamedTemporaryFile(delete=False) as file:
```

Comment:
Generally speaking, it's better not to leak temporary files (which is what `delete=False` does, unless you explicitly delete it later). What is the reasoning for using `delete=false`?

Reply:
I might be wrong, but if I don't add `delete=False`, it complains it cannot find the tmp file on delete. I was thinking maybe `with` context and `tempfile` both try to delete the file when exiting the context, as a result, one of them hits the error. But let me double check if that is the case, and will ad a comment if yes.

Reply:
I would be extremely surprised if that were the case. Here's a simple test:

```
macbook-pro-116:~ ezyang$ cat test.py
import tempfile

with tempfile.NamedTemporaryFile() as f:
    pass
macbook-pro-116:~ ezyang$ python test.py
macbook-pro-116:~ ezyang$ 
```